dali/image/tiff.cc

```diff
@@ -0,0 +1,69 @@
+#include "tiff.h"
+
+namespace dali {
+
+namespace {
+
+cv::Mat DecodeTiff(const unsigned char *tiff, int size, DALIImageType image_type) {
+    std::vector<char> buf(tiff, tiff + size);
+    return cv::imdecode(buf, image_type == DALI_GRAY ? 0 : 1);
+}
+
+
+} // namespace
+
+bool CheckIsTiff(const unsigned char *tiff) {
+    assert(tiff);
+
+    std::vector<int> header_intel = {77, 77, 0, 42};
+    std::vector<int> header_motorola = {73, 73, 42, 0};
+
+    auto check_header = [&](const std::vector<int> &header) -> bool {
+        for (unsigned int i = 0; i < header.size(); i++) {
+            if (tiff[i] != header[i]) {
+                return false;
+            }
+        }
+        return true;
+    };
+
+    return check_header(header_intel) || check_header(header_motorola);
+}
+
+
+DALIError_t GetTiffImageDims(const unsigned char *tiff, int size, int *h, int *w) {
+    assert(CheckIsTiff(tiff));
+    auto tiff_mat = DecodeTiff(tiff, size, DALI_RGB);
```

Comment:
No, this way we decode image twice, once to just get its dimensions, and second time when we do fallback to host from the nvJpeg.
We need to parse header on our own to extract dimensions.

Reply:
I'm aware of this. However, I've done some research on this subject and all of libraries, that handle tiff images, perform entire decoding, ever operations such as getting image dimmensions. I suppose, it's because tiff header is not as simple as png or bmp - sometimes it requires a lot of processing, because header data can be split across whole file.

Taking large amount of work needed to extract manually image dimmensions without explicit guarantee, that it will be in fact quicker, that using OpenCV to decode image, I decided to do it this way. If you think, that it's bad approach, I will of course do it manually.

I would also like to add a note here: It seems, that `image` module could be nicely coverted to more object-oriented approach, what you think about this?

Reply:
Done